feat: add methods for mutating events, editions, slots, snapshots

rs-amp opened the pull request:
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Tests (`npm run test`) for the changes have been added (for bug fixes / features)
- [x] Docs (`npm run doc`) have been reviewed and added / updated if needed (for bug fixes / features)
- [x] PR title and git commit messages conform to the [conventionalcommits](https://www.conventionalcommits.org/en/v1.0.0/) specification


## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behaviour?
Manangement SDK is missing a few methods for mutating events and related resources, as well as a few common fields.

## What is the new behaviour?
This PR adds methods for mutating events, editions, slots and snapshots:

- Snapshot Create from Hub
- Event Update
- Edition Update
- Edition Slot Create
- Edition Slot Content Update
- Missing fields on Snapshot, Edition Slot

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This PR introduces the first method that uses the PUT method - the `safe-update-content` method on EditionSlots. I've made a few changes to support this in both the HalClient and the tests.

paulcrussell commented on src/lib/model/Hub.ts:
```diff
@@ -281,6 +283,20 @@ export class Hub extends HalResource {
 
+    snapshots: {
+      /**
+       * Creates a Webhook inside this Hub
```
Change this comment

PikinerLiliya replied:
+

rs-amp commented on src/lib/hal/services/HalClient.ts:
```diff
@@ -35,7 +35,8 @@ export interface HalClient {
     resourceConstructor: HalResourceConstructor<T>,
-    method: HttpMethod.POST | HttpMethod.PATCH
+    method: HttpMethod.POST | HttpMethod.PATCH | HttpMethod.PUT,
+    returnError: boolean
```
The point is that the data in the error is meaningful. When the data is put into the HttpError, it isn't parsed as an object, it's just put into the message string. In the case of edition schedule, the return value is a status object containing an array of warnings/errors, each with codes, messages and overlaps of their own.

easen-amp replied:
This could be modelled using an Exception, the validation errors that DC throw all follow the same pattern --> https://amplience.com/docs/api/dynamic-content/management/index.html#section/Usage/Error-Response-Format

rs-amp replied:
I could also remove this functionality for now and just leave it up to the user to parse/interpret whatever's in `HttpError.response.data.errors[...]` with the interfaces left exported. That might be a better way to get these changes in without any larger scale changes to the error exception type. What do you think?

easen-amp replied:
Is the `HttpError` from underlying Axios lib? If so, I think we should create our own HTTP Error Exception class for this.

Responses & Exceptions form part of the 'API contract' SDK & their users, we should try to minimise the use 3rd party libs from these 'contracts'.

rs-amp replied:
The HttpError is provided by us, but it's safe to assume that people have built tools catching these exceptions already, so any changes should be built on top of these error types if possible. I'm just asking for now though, if it would make sense for me to just leave this change out and parse the errors directly from the data in my pending changes on the `dc-cli` side.

rs-amp commented:
I've changed things so that edition schedule throws the HttpError like the other methods, and it's up to the user to handle what's in the `response.data` themselves for now. I've removed the `returnError` functionality from the HttpClient as a result.

`performActionWithoutResourceResponse` has been expanded to support passing query string parameters and data, as `schedule` now doesn't have a return type.
